File: solutions/cpp/collatz-conjecture/1/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include<stdexcept>

namespace collatz_conjecture {

    unsigned int collatz_map(unsigned int num){
        int N;
        if (!(num % 2)) 
            N = num/2;
        else 
            N = 3*num + 1;
        return N;
    }

    int steps(int num){
        int counter = 0;
        if (num<=0)
            throw std::domain_error("Number must be strictly postive.");
        while(num > 1){
            num = collatz_map(num);
            counter++;
        }
        return counter;    
        
    }
}  // namespace collatz_conjecture

File: solutions/fortran/collatz-conjecture/1/collatz_conjecture.f90

module collatz_conjecture
  implicit none
contains

  integer function collatz_map(n)
    integer :: n
    if (mod(n,2) == 0) then
      n = n/2
    else 
      n = 3*n + 1
   end if
    collatz_map = n
  end function collatz_map

  integer function steps(i)
    integer, intent(in) :: i
    integer ::  j, stp_count
    stp_count = -1
     j = i
    if (j>0) then
       stp_count = 0
       do while (j /= 1) 
          j = collatz_map(j)
          stp_count = stp_count + 1
       end do
    end if
    steps = stp_count
  end function steps

end module collatz_conjecture
